src/file.cr

```diff
@@ -458,40 +458,45 @@ class File < IO::FileDescriptor
   #
   # The pattern syntax is similar to shell filename globbing. It may contain the following metacharacters:
   #
-  # * `*` matches an unlimited number of arbitrary characters excluding `/`.
+  # * `*` matches an unlimited number of arbitrary characters, excluding any directory separators.
   #   * `"*"` matches all regular files.
   #   * `"c*"` matches all files beginning with `c`.
   #   * `"*c"` matches all files ending with `c`.
   #   * `"*c*"` matches all files that have `c` in them (including at the beginning or end).
   # * `**` matches directories recursively if followed by `/`.
   #   If this path segment contains any other characters, it is the same as the usual `*`.
-  # * `?` matches any one character excluding `/`.
+  # * `?` matches one arbitrary character, excluding any directory separators.
   # * character sets:
-  #   * `[abc]` matches any one of these character.
+  #   * `[abc]` matches any one of these characters.
   #   * `[^abc]` matches any one character other than these.
   #   * `[a-z]` matches any one character in the range.
   # * `{a,b}` matches subpattern `a` or `b`.
   # * `\\` escapes the next character.
   #
-  # NOTE: Only `/` is recognized as path separator in both *pattern* and *path*.
+  # If *path* is a `Path`, all directory separators supported by *path* are
+  # recognized, according to the path's kind. If *path* is a `String`, only `/`
+  # is considered a directory separator.
+  #
+  # NOTE: Only `/` in *pattern* matches directory separators in *path*.
   def self.match?(pattern : String, path : Path | String) : Bool
     expanded_patterns = [] of String
     File.expand_brace_pattern(pattern, expanded_patterns)
 
     expanded_patterns.each do |expanded_pattern|
-      return true if match_single_pattern(expanded_pattern, path.to_s)
+      return true if match_single_pattern(expanded_pattern, path)
     end
     false
   end
 
-  private def self.match_single_pattern(pattern : String, path : String)
+  private def self.match_single_pattern(pattern : String, path : Path | String)
```

Comment:
This is a lower-level helper method, and it feels bit iffy that it handles these different input types.
I'm wondering if it would be helpful to resolve the union which could happen in two ways:
* receive only `Path` and have `.match?` transform `String` into `Path.posix`
* receive only `String` with an additional `separators` parameter which is resolved in `.match?`

WDYT?

Reply:
It could be moved to the caller even; there is no need to do this once per pattern when there are brace expansions

Reply:
It could even be moved into the caller entirely; there is no need to handle `Path`s once per pattern resulting from brace expansion

Reply:
Hm I don't know. What would the benefit be?
Calling `Path.separators` isn't very costly. I'd bet it gets inlined anyway.